File: C# Unity Build/Dragon_Raid/Assets/Scripts/EnemyAi.cs
using UnityEngine;
using TMPro;
using Unity.VisualScripting;


public class EnemyAI : MonoBehaviour
{
    public EnemyData baseStats;
    public int currentHealth;
    public int currentAttack;
    public int currentDefense;
    public int currentLuck;
    public int currentPower; // Enemy basic attack power

    public CombatManager combatManager;
    private PlayerController playerTarget;

    [Header("UI")]
    public TextMeshProUGUI healthText; // Drag your new text object here
    
    public int barLength = 10;
    public char fillChar = '█';
    public char emptyChar = '░';
    // -------------------------

    public void Setup(EnemyData data, CombatManager manager, PlayerController player)
    {
        baseStats = data;
        combatManager = manager;
        playerTarget = player;

        currentHealth = baseStats.maxHealth;
        currentAttack = baseStats.attack;
        currentDefense = baseStats.defense;
        currentLuck = baseStats.luck;
        currentPower = baseStats.power;

        UpdateUI();
    }

    public void UpdateUI()
    {
        if (playerTarget != null && healthText != null && combatManager != null)
        {
            healthText.text = combatManager.GenerateBarString(baseStats.enemyName,currentHealth, baseStats.maxHealth);
        }
    }

    // This is the enemy's basic attack
    public void TakeTurn()
    {
        if (playerTarget != null)
        {
            Debug.Log(baseStats.enemyName + " attacks the player!");



            // The enemy just uses its basic stats for its turn
            playerTarget.TakeDamage(currentAttack, currentLuck, currentPower);
        }
    }

    public void TakeDamage(int attackerAttack, int attackerLuck, int attackerPower)
    {

        float randMin = combatManager.randMinMultiplier;
        float randMax = combatManager.randMaxMultiplier;
        float atkPotency = combatManager.attackPotency;
        float defDivisor = combatManager.defenseDivisor;
        float defPotency = combatManager.defensePotency;
        float defPenalty = combatManager.defensePenaltyMultiplier;
        float defScalar = combatManager.defensePenaltyScalar;
        float defToggle = combatManager.defensePenaltyToggle;
        float flatBonus = combatManager.flatDamageBonus;

        float critChancePerLuck = combatManager.lucktoCrit;
        float critMultiplier = combatManager.critDamageMultiplier;


        // Formula stats
        float ATK = attackerAttack;
        float PWR = attackerPower;
        float DEF = this.currentDefense; // This is the enemy's defense

        float randValue = Random.Range(randMin * PWR, randMax * PWR);

        if (defDivisor == 0) defDivisor = 1f;

        float mainMultiplier = (ATK * atkPotency) / (DEF * defDivisor);
        float penalty = (DEF * defPotency * defPenalty) * defScalar * defToggle;

        int baseDamage = Mathf.FloorToInt(randValue * (mainMultiplier - penalty) + flatBonus);

        float critChance = attackerLuck * critChancePerLuck;
        float critRoll = Random.Range(0f, 100f);

        int finalDamage;
        if (critRoll < critChance)
        {
            finalDamage = (int)(Mathf.FloorToInt(baseDamage * critMultiplier));
            Debug.Log("Critical Hit!");
        }
        else
        {
            finalDamage = baseDamage;
        }
        if (finalDamage < 1) finalDamage = 1;

        currentHealth -= finalDamage;
        Debug.Log(baseStats.enemyName + " takes " + finalDamage + " damage.");

        UpdateUI();

        if (currentHealth <= 0)
        {
            Die();
        }


    }



    void Die()
    {
        combatManager.EnemyDied(this.gameObject);
        Destroy(this.gameObject);
    }

    // console command to immediately kill this enemy
    public void ForceKill()
    {
        currentHealth = 0;
        Die();
    }
}

File: DragonRR/DragonRR/src/commonMain/kotlin/samples/mygame/enemy/enemyAI.kt
package com.yourgame.enemy
import korlibs.korge.view.*
import korlibs.korge.view.Text
import com.yourgame.combat.CombatManager
import com.yourgame.player.PlayerController
import com.yourgame.data.EnemyData
import kotlin.math.floor
import kotlin.random.Random

class EnemyAI(
    var combatManager: CombatManager? = null,
    var playerTarget: PlayerController? = null
) {
    var view: View? = null

    lateinit var baseStats: EnemyData

    var currentHealth: Int = 0
    var currentAttack: Int = 0
    var currentDefense: Int = 0
    var currentLuck: Int = 0
    var currentPower: Int = 0  // Enemy basic attack power

    // UI
    var healthText: Text? = null

    var barLength: Int = 10
    var fillChar: Char = '█'
    var emptyChar: Char = '░'
    // -------------------------

    fun setup(data: EnemyData, manager: CombatManager, player: PlayerController) {
        baseStats = data
        combatManager = manager
        playerTarget = player

        currentHealth = baseStats.maxHealth
        currentAttack = baseStats.attack
        currentDefense = baseStats.defense
        currentLuck = baseStats.luck
        currentPower = baseStats.power

        updateUI()
    }

    fun attachView(enemyView: View) {
        view = enemyView
        enemyView.name = baseStats.enemyName
    }

    fun updateUI() {
        val manager = combatManager ?: return
        val player = playerTarget ?: return
        healthText?.text = manager.generateBarString(baseStats.enemyName, currentHealth, baseStats.maxHealth)
    }

    //enemy's basic attack
    fun takeTurn() {
        val player = playerTarget ?: return
        println("${baseStats.enemyName} attacks the player!")

        //enemy uses its basic stats for its turn
        player.takeDamage(currentAttack, currentLuck, currentPower)
    }

    fun takeDamage(attackerAttack: Int, attackerLuck: Int, attackerPower: Int) {
        val manager = combatManager ?: return

        val randMin = manager.randMinMultiplier
        val randMax = manager.randMaxMultiplier
        val atkPotency = manager.attackPotency
        val defDivisor = manager.defenseDivisor.coerceAtLeast(1f)
        val defPotency = manager.defensePotency
        val defPenalty = manager.defensePenaltyMultiplier
        val defScalar = manager.defensePenaltyScalar
        val defToggle = manager.defensePenaltyToggle
        val flatBonus = manager.flatDamageBonus

        val critChancePerLuck = manager.lucktoCrit
        val critMultiplier = manager.critDamageMultiplier

        val ATK = attackerAttack
        val PWR = attackerPower
        val DEF = currentDefense  // This is the enemy's defense

        val randValue = Random.nextFloat() * (randMax - randMin) * PWR + randMin * PWR

        val mainMultiplier = (ATK * atkPotency) / (DEF * defDivisor)
        val penalty = (DEF * defPotency * defPenalty) * defScalar * defToggle

        val baseDamage = floor((randValue * (mainMultiplier - penalty) + flatBonus).toDouble()).toInt()

        val critChance = attackerLuck * critChancePerLuck
        val critRoll = Random.nextFloat() * 100f

        val finalDamage = if (critRoll < critChance) {
            println("Critical Hit!")
            floor((baseDamage.toDouble() * critMultiplier.toDouble())).toInt()
        } else {
            baseDamage
        }.coerceAtLeast(1)

        currentHealth -= finalDamage
        println("${baseStats.enemyName} takes $finalDamage damage.")

        updateUI()

        if (currentHealth <= 0) {
            die()
        }
    }

    private fun die() {
        // TODO: notify manager if needed
        //remove the enemy view
    }
    //console command to immediately kill this enemy
    fun forceKill() {
        currentHealth = 0
        die()
    }
}
